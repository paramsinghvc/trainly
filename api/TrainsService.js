"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrainsService = exports.Trains = exports.BASE_URL = void 0;
const axios_1 = __importDefault(require("axios"));
const xml2js_1 = require("xml2js");
const util_1 = require("util");
const lodash_1 = require("lodash");
const toCamelCase = (str) => str.charAt(0).toLocaleLowerCase() + str.slice(1);
const xml2JSON = (0, util_1.promisify)(xml2js_1.parseString);
exports.BASE_URL = 'https://lite.realtime.nationalrail.co.uk/OpenLDBWS/ldb11.asmx';
const ACCESS_TOKEN = process.env.ACCESS_TOKEN;
var Trains;
(function (Trains) {
    let Operation;
    (function (Operation) {
        Operation["GetDepartureBoard"] = "GetDepartureBoard";
        Operation["GetDepBoardWithDetails"] = "GetDepBoardWithDetails";
        Operation["GetArrivalBoard"] = "GetArrivalBoard";
        Operation["GetArrBoardWithDetails"] = "GetArrBoardWithDetails";
        Operation["GetArrivalDepartureBoard"] = "GetArrivalDepartureBoard";
        Operation["GetArrDepBoardWithDetails"] = "GetArrDepBoardWithDetails";
        Operation["GetNextDepartures"] = "GetNextDepartures";
        Operation["GetNextDeparturesWithDetails"] = "GetNextDeparturesWithDetails";
        Operation["GetFastestDepartures"] = "GetFastestDepartures";
        Operation["GetFastestDeparturesWithDetails"] = "GetFastestDeparturesWithDetails";
        Operation["GetServiceDetails"] = "GetServiceDetails";
    })(Operation = Trains.Operation || (Trains.Operation = {}));
})(Trains = exports.Trains || (exports.Trains = {}));
class TrainsService {
    constructor() {
        this.postArrayProcessor = (obj) => {
            objProcessor(obj, {
                callingPointList: (val) => getArrayValue(val === null || val === void 0 ? void 0 : val.callingPoint),
                trainServices: (trainServices) => getArrayValue(trainServices === null || trainServices === void 0 ? void 0 : trainServices.service),
            });
            return obj;
        };
    }
    constructXMLFromObject(obj) {
        return Object.entries(obj).reduce((acc, [key, value]) => {
            const val = key === 'filterList' ? value.reduce((acc, v) => acc + `<crs>${v}</crs>`, '') : value;
            return acc + (typeof val !== 'undefined' ? `<${key}>${val}</${key}>` : ''); //
        }, '');
    }
    constructXMLBody(operationName, payload) {
        const body = this.constructXMLFromObject(payload);
        return `<?xml version="1.0"?>
		<Envelope xmlns="http://www.w3.org/2003/05/soap-envelope">
			<Header>
				<AccessToken xmlns="http://thalesgroup.com/RTTI/2013-11-28/Token/types">
					<TokenValue>${ACCESS_TOKEN}</TokenValue>
				</AccessToken>
			</Header>
			<Body>
				<${operationName}Request xmlns="http://thalesgroup.com/RTTI/2017-10-01/ldb/">
					${body}
				</${operationName}Request>
			</Body>
		</Envelope>`;
    }
    fetchData(operationName, payload) {
        return __awaiter(this, void 0, void 0, function* () {
            const xmlPayload = this.constructXMLBody(operationName, payload);
            console.log(xmlPayload);
            const result = yield (0, axios_1.default)({
                method: 'POST',
                url: exports.BASE_URL,
                headers: {
                    'Content-Type': 'text/xml',
                },
                data: xmlPayload,
            });
            const data = yield this.parseResponse(result.data, operationName);
            return data;
        });
    }
    parseResponse(result, operationName) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield xml2JSON(result, {
                trim: true,
                ignoreAttrs: true,
                explicitArray: false,
                tagNameProcessors: [(name) => (name.includes(':') ? name.split(':')[1] : name)],
            })
                .then((data) => (0, lodash_1.get)(data, ['Envelope', 'Body', `${operationName}Response`]))
                .then(this.postArrayProcessor);
        });
    }
}
exports.TrainsService = TrainsService;
const getArrayValue = (val) => (val ? ((0, lodash_1.isArray)(val) ? val : [val]) : null);
function objProcessor(obj, config) {
    const configKeys = Object.keys(config);
    Object.entries(obj).forEach(([key, val]) => {
        if (configKeys.includes(key)) {
            obj[key] = config[key](val);
        }
        if ((0, lodash_1.isObjectLike)(obj[key])) {
            objProcessor(obj[key], config);
        }
    });
    return obj;
}
