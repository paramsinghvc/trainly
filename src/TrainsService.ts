import axios from 'axios';
import { parseString, convertableToString, ParserOptions, Builder } from 'xml2js';
import { promisify } from 'util';
import { get } from 'lodash';

type ParseString = (
  str: convertableToString,
  options: ParserOptions,
  callback: (err: Error | null, result: any) => void
) => void;

const xml2JSON = promisify(parseString as ParseString);

const BASE_URL = 'https://lite.realtime.nationalrail.co.uk/OpenLDBWS/ldb9.asmx';

const ACCESS_TOKEN = process.env.ACCESS_TOKEN;

console.log({ ACCESS_TOKEN });
export namespace Trains {
  export enum Operation {
    GetDepartureBoard = 'GetDepartureBoard',
    GetDepBoardWithDetails = 'GetDepBoardWithDetails',
    GetArrivalBoard = 'GetArrivalBoard',
    GetArrBoardWithDetails = 'GetArrBoardWithDetails',
    GetArrivalDepartureBoard = 'GetArrivalDepartureBoard',
    GetArrDepBoardWithDetails = 'GetArrDepBoardWithDetails',
    GetNextDepartures = 'GetNextDepartures',
    GetNextDeparturesWithDetails = 'GetNextDeparturesWithDetails',
    GetFastestDepartures = 'GetFastestDepartures',
    GetFastestDeparturesWithDetails = 'GetFastestDeparturesWithDetails',
    GetServiceDetails = 'GetServiceDetails',
  }

  // Data for Board Calls
  export type Board = {
    numRows?: number;
    crs: string;
    filterCrs?: string;
    filterType?: string;
    timeOffset?: number;
    timeWindow?: number;
  };

  // Data for Departure Calls
  export type Departure = {
    crs: string;
    filterList?: string[];
    timeOffset?: number;
    timeWindow?: number;
  };

  // Data for the GetServiceDetails call
  export type Service = {
    serviceID: string;
  };

  export type Payload = Board | Departure | Service;
}

export class TrainsService {
  constructXMLFromObject(obj: { [k: string]: any }) {
    return Object.entries(obj).reduce((acc, [key, value]) => {
      const val = key === 'filterList' ? value.reduce((acc: string, v: string) => acc + `<crs>${v}</crs>`, '') : value;
      return acc + `<${key}>${val}</${key}>`;
    }, '');
  }

  constructXMLBody(operationName: Trains.Operation, payload: Trains.Payload) {
    const body = this.constructXMLFromObject(payload);
    return `<?xml version="1.0"?>
		<Envelope xmlns="http://www.w3.org/2003/05/soap-envelope">
			<Header>
				<AccessToken xmlns="http://thalesgroup.com/RTTI/2013-11-28/Token/types">
					<TokenValue>${ACCESS_TOKEN}</TokenValue>
				</AccessToken>
			</Header>
			<Body>
				<${operationName}Request xmlns="http://thalesgroup.com/RTTI/2016-02-16/ldb/">
					${body}
				</${operationName}Request>
			</Body>
		</Envelope>`;
  }

  async fetchData(operationName: Trains.Operation, payload: Trains.Payload) {
    const xmlPayload = this.constructXMLBody(operationName, payload);
    const result = await axios({
      method: 'POST',
      url: BASE_URL,
      headers: {
        'Content-Type': 'text/xml',
      },
      data: xmlPayload,
    });

    const data = await this.parseResponse(result.data, operationName);

    console.log(data);
    return data;
  }

  async parseResponse(result: string, operationName: Trains.Operation) {
    return await xml2JSON(result, {
      trim: true,
      ignoreAttrs: true,
      explicitArray: false,
      tagNameProcessors: [(name) => (name.includes(':') ? name.split(':')[1] : name)],
    }).then((data) => get(data, ['Envelope', 'Body', `${operationName}Response`]));
  }
}
