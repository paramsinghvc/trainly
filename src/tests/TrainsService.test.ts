import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { BASE_URL, Trains, TrainsService } from '../TrainsService';
import { MOCK_APP_DEP_WITH_DETAILS_FROM_ONLY } from './mockXMLResponse';

let axiosMock = new MockAdapter(axios);

describe('Trains Service', () => {
  let trainsService: TrainsService;
  beforeAll(() => {
    trainsService = new TrainsService();
  });

  it('should fetch data for GetArrDepWithDetails', async () => {
    axiosMock.onPost(BASE_URL).reply(200, MOCK_APP_DEP_WITH_DETAILS_FROM_ONLY);
    const data = await trainsService.fetchData(Trains.Operation.GetArrDepBoardWithDetails, {
      crs: 'KGX',
    });

    expect(data).toMatchSnapshot();
  });
});
