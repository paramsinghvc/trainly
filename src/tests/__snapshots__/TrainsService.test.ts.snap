// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trains Service should fetch data for GetArrDepWithDetails 1`] = `
{
  "GetStationBoardResult": {
    "crs": "KGX",
    "generatedAt": "2023-04-16T00:17:56.0709132+01:00",
    "locationName": "London Kings Cross",
    "platformAvailable": "true",
    "trainServices": [
      {
        "destination": {
          "location": {
            "crs": "KGX",
            "locationName": "London Kings Cross",
          },
        },
        "eta": "On time",
        "length": "8",
        "operator": "Thameslink",
        "operatorCode": "TL",
        "origin": {
          "location": {
            "crs": "PBO",
            "locationName": "Peterborough",
          },
        },
        "platform": "9",
        "previousCallingPoints": {
          "callingPointList": [
            {
              "at": "On time",
              "crs": "PBO",
              "length": "8",
              "locationName": "Peterborough",
              "st": "22:54",
            },
            {
              "at": "On time",
              "crs": "HUN",
              "length": "8",
              "locationName": "Huntingdon",
              "st": "23:10",
            },
            {
              "at": "On time",
              "crs": "SNO",
              "length": "8",
              "locationName": "St Neots",
              "st": "23:18",
            },
            {
              "at": "On time",
              "crs": "SDY",
              "length": "8",
              "locationName": "Sandy",
              "st": "23:26",
            },
            {
              "at": "On time",
              "crs": "BIW",
              "length": "8",
              "locationName": "Biggleswade",
              "st": "23:30",
            },
            {
              "at": "On time",
              "crs": "ARL",
              "length": "8",
              "locationName": "Arlesey",
              "st": "23:35",
            },
            {
              "at": "On time",
              "crs": "HIT",
              "length": "8",
              "locationName": "Hitchin",
              "st": "23:44",
            },
            {
              "at": "On time",
              "crs": "SVG",
              "length": "8",
              "locationName": "Stevenage",
              "st": "23:50",
            },
            {
              "at": "On time",
              "crs": "FPK",
              "length": "8",
              "locationName": "Finsbury Park",
              "st": "00:10",
            },
          ],
        },
        "serviceID": "2664531KNGX____",
        "serviceType": "train",
        "sta": "00:18",
      },
      {
        "destination": {
          "location": {
            "crs": "RYS",
            "locationName": "Royston",
          },
        },
        "etd": "On time",
        "length": "8",
        "operator": "Great Northern",
        "operatorCode": "GN",
        "origin": {
          "location": {
            "crs": "KGX",
            "locationName": "London Kings Cross",
          },
        },
        "platform": "9",
        "rsid": "GN017200",
        "serviceID": "2671400KNGX____",
        "serviceType": "train",
        "std": "00:33",
        "subsequentCallingPoints": {
          "callingPointList": [
            {
              "crs": "FPK",
              "et": "On time",
              "length": "8",
              "locationName": "Finsbury Park",
              "st": "00:38",
            },
            {
              "crs": "SVG",
              "et": "On time",
              "length": "8",
              "locationName": "Stevenage",
              "st": "01:03",
            },
            {
              "crs": "HIT",
              "et": "On time",
              "length": "8",
              "locationName": "Hitchin",
              "st": "01:12",
            },
            {
              "crs": "LET",
              "et": "On time",
              "length": "8",
              "locationName": "Letchworth Garden City",
              "st": "01:18",
            },
            {
              "crs": "BDK",
              "et": "On time",
              "length": "8",
              "locationName": "Baldock",
              "st": "01:21",
            },
            {
              "crs": "RYS",
              "et": "On time",
              "length": "8",
              "locationName": "Royston",
              "st": "01:29",
            },
          ],
        },
      },
      {
        "delayReason": "This train has been delayed by trespassers on the railway earlier today",
        "destination": {
          "location": {
            "crs": "KGX",
            "locationName": "London Kings Cross",
          },
        },
        "eta": "On time",
        "length": "8",
        "operator": "Thameslink",
        "operatorCode": "TL",
        "origin": {
          "location": {
            "crs": "PBO",
            "locationName": "Peterborough",
          },
        },
        "previousCallingPoints": {
          "callingPointList": [
            {
              "at": "23:25",
              "crs": "PBO",
              "length": "8",
              "locationName": "Peterborough",
              "st": "23:20",
            },
            {
              "at": "23:39",
              "crs": "HUN",
              "length": "8",
              "locationName": "Huntingdon",
              "st": "23:35",
            },
            {
              "at": "23:47",
              "crs": "SNO",
              "length": "8",
              "locationName": "St Neots",
              "st": "23:43",
            },
            {
              "at": "23:55",
              "crs": "SDY",
              "length": "8",
              "locationName": "Sandy",
              "st": "23:50",
            },
            {
              "at": "23:59",
              "crs": "BIW",
              "length": "8",
              "locationName": "Biggleswade",
              "st": "23:54",
            },
            {
              "at": "00:04",
              "crs": "ARL",
              "length": "8",
              "locationName": "Arlesey",
              "st": "23:59",
            },
            {
              "at": "00:10",
              "crs": "HIT",
              "length": "8",
              "locationName": "Hitchin",
              "st": "00:07",
            },
            {
              "at": "00:16",
              "crs": "SVG",
              "length": "8",
              "locationName": "Stevenage",
              "st": "00:13",
            },
            {
              "crs": "FPK",
              "et": "On time",
              "length": "8",
              "locationName": "Finsbury Park",
              "st": "00:40",
            },
          ],
        },
        "rsid": "TL297100",
        "serviceID": "2643106KNGX____",
        "serviceType": "train",
        "sta": "00:46",
      },
      {
        "cancelReason": "This train has been cancelled because of more trains than usual needing repairs at the same time",
        "destination": {
          "location": {
            "crs": "KGX",
            "locationName": "London Kings Cross",
          },
        },
        "eta": "Cancelled",
        "isCancelled": "true",
        "operator": "Thameslink",
        "operatorCode": "TL",
        "origin": {
          "location": {
            "crs": "RYS",
            "locationName": "Royston",
          },
        },
        "previousCallingPoints": {
          "callingPointList": [
            {
              "crs": "RYS",
              "et": "Cancelled",
              "locationName": "Royston",
              "st": "23:55",
            },
            {
              "crs": "AWM",
              "et": "Cancelled",
              "locationName": "Ashwell & Morden",
              "st": "23:59",
            },
            {
              "crs": "BDK",
              "et": "Cancelled",
              "locationName": "Baldock",
              "st": "00:04",
            },
            {
              "crs": "LET",
              "et": "Cancelled",
              "locationName": "Letchworth Garden City",
              "st": "00:08",
            },
            {
              "crs": "HIT",
              "et": "Cancelled",
              "locationName": "Hitchin",
              "st": "00:12",
            },
            {
              "crs": "SVG",
              "et": "Cancelled",
              "locationName": "Stevenage",
              "st": "00:18",
            },
            {
              "crs": "KBW",
              "et": "Cancelled",
              "locationName": "Knebworth",
              "st": "00:21",
            },
            {
              "crs": "WLW",
              "et": "Cancelled",
              "locationName": "Welwyn North",
              "st": "00:25",
            },
            {
              "crs": "WGC",
              "et": "Cancelled",
              "locationName": "Welwyn Garden City",
              "st": "00:30",
            },
            {
              "crs": "HAT",
              "et": "Cancelled",
              "locationName": "Hatfield",
              "st": "00:34",
            },
            {
              "crs": "PBR",
              "et": "Cancelled",
              "locationName": "Potters Bar",
              "st": "00:39",
            },
            {
              "crs": "FPK",
              "et": "Cancelled",
              "locationName": "Finsbury Park",
              "st": "00:51",
            },
          ],
        },
        "rsid": "TL061100",
        "serviceID": "2661513KNGX____",
        "serviceType": "train",
        "sta": "00:58",
      },
      {
        "cancelReason": "This train has been cancelled because of more trains than usual needing repairs at the same time",
        "delayReason": "This train has been delayed by more trains than usual needing repairs at the same time",
        "destination": {
          "location": {
            "crs": "KGX",
            "locationName": "London Kings Cross",
          },
        },
        "eta": "On time",
        "length": "8",
        "operator": "Thameslink",
        "operatorCode": "TL",
        "origin": {
          "location": {
            "crs": "RYS",
            "locationName": "Royston",
          },
        },
        "previousCallingPoints": {
          "callingPointList": [
            {
              "at": "On time",
              "crs": "RYS",
              "length": "8",
              "locationName": "Royston",
              "st": "00:09",
            },
            {
              "at": "On time",
              "crs": "AWM",
              "length": "8",
              "locationName": "Ashwell & Morden",
              "st": "00:15",
            },
            {
              "crs": "BDK",
              "et": "00:19",
              "length": "8",
              "locationName": "Baldock",
              "st": "00:17",
            },
            {
              "crs": "LET",
              "et": "On time",
              "length": "8",
              "locationName": "Letchworth Garden City",
              "st": "00:21",
            },
            {
              "crs": "HIT",
              "et": "On time",
              "length": "8",
              "locationName": "Hitchin",
              "st": "00:28",
            },
            {
              "crs": "SVG",
              "et": "On time",
              "length": "8",
              "locationName": "Stevenage",
              "st": "00:34",
            },
            {
              "crs": "KBW",
              "et": "On time",
              "length": "8",
              "locationName": "Knebworth",
              "st": "00:36",
            },
            {
              "crs": "WLW",
              "et": "On time",
              "length": "8",
              "locationName": "Welwyn North",
              "st": "00:38",
            },
            {
              "crs": "WGC",
              "et": "On time",
              "length": "8",
              "locationName": "Welwyn Garden City",
              "st": "00:41",
            },
            {
              "crs": "HAT",
              "et": "On time",
              "length": "8",
              "locationName": "Hatfield",
              "st": "00:42",
            },
            {
              "crs": "PBR",
              "et": "On time",
              "length": "8",
              "locationName": "Potters Bar",
              "st": "00:46",
            },
            {
              "crs": "FPK",
              "et": "On time",
              "length": "8",
              "locationName": "Finsbury Park",
              "st": "01:00",
            },
          ],
        },
        "rsid": "TL094800",
        "serviceID": "2671490KNGX____",
        "serviceType": "train",
        "sta": "01:06",
      },
      {
        "destination": {
          "location": {
            "crs": "HFN",
            "locationName": "Hertford North",
          },
        },
        "etd": "On time",
        "length": "6",
        "operator": "Great Northern",
        "operatorCode": "GN",
        "origin": {
          "location": {
            "crs": "KGX",
            "locationName": "London Kings Cross",
          },
        },
        "rsid": "GN008700",
        "serviceID": "2671502KNGX____",
        "serviceType": "train",
        "std": "01:10",
        "subsequentCallingPoints": {
          "callingPointList": [
            {
              "crs": "FPK",
              "et": "On time",
              "length": "6",
              "locationName": "Finsbury Park",
              "st": "01:15",
            },
            {
              "crs": "HGY",
              "et": "On time",
              "length": "6",
              "locationName": "Harringay",
              "st": "01:18",
            },
            {
              "crs": "HRN",
              "et": "On time",
              "length": "6",
              "locationName": "Hornsey",
              "st": "01:20",
            },
            {
              "crs": "AAP",
              "et": "On time",
              "length": "6",
              "locationName": "Alexandra Palace",
              "st": "01:23",
            },
            {
              "crs": "BOP",
              "et": "On time",
              "length": "6",
              "locationName": "Bowes Park",
              "st": "01:26",
            },
            {
              "crs": "PAL",
              "et": "On time",
              "length": "6",
              "locationName": "Palmers Green",
              "st": "01:28",
            },
            {
              "crs": "WIH",
              "et": "On time",
              "length": "6",
              "locationName": "Winchmore Hill",
              "st": "01:31",
            },
            {
              "crs": "GPK",
              "et": "On time",
              "length": "6",
              "locationName": "Grange Park",
              "st": "01:33",
            },
            {
              "crs": "ENC",
              "et": "On time",
              "length": "6",
              "locationName": "Enfield Chase",
              "st": "01:35",
            },
            {
              "crs": "GDH",
              "et": "On time",
              "length": "6",
              "locationName": "Gordon Hill",
              "st": "01:37",
            },
            {
              "crs": "CUF",
              "et": "On time",
              "length": "6",
              "locationName": "Cuffley",
              "st": "01:41",
            },
            {
              "crs": "HFN",
              "et": "On time",
              "length": "6",
              "locationName": "Hertford North",
              "st": "01:49",
            },
          ],
        },
      },
      {
        "destination": {
          "location": {
            "crs": "WGC",
            "locationName": "Welwyn Garden City",
          },
        },
        "etd": "On time",
        "length": "6",
        "operator": "Great Northern",
        "operatorCode": "GN",
        "origin": {
          "location": {
            "crs": "KGX",
            "locationName": "London Kings Cross",
          },
        },
        "rsid": "GN059400",
        "serviceID": "2671405KNGX____",
        "serviceType": "train",
        "std": "01:40",
        "subsequentCallingPoints": {
          "callingPointList": [
            {
              "crs": "FPK",
              "et": "On time",
              "length": "6",
              "locationName": "Finsbury Park",
              "st": "01:45",
            },
            {
              "crs": "AAP",
              "et": "On time",
              "length": "6",
              "locationName": "Alexandra Palace",
              "st": "01:53",
            },
            {
              "crs": "NSG",
              "et": "On time",
              "length": "6",
              "locationName": "New Southgate",
              "st": "01:56",
            },
            {
              "crs": "OKL",
              "et": "On time",
              "length": "6",
              "locationName": "Oakleigh Park",
              "st": "01:59",
            },
            {
              "crs": "NBA",
              "et": "On time",
              "length": "6",
              "locationName": "New Barnet",
              "st": "02:01",
            },
            {
              "crs": "PBR",
              "et": "On time",
              "length": "6",
              "locationName": "Potters Bar",
              "st": "02:07",
            },
            {
              "crs": "HAT",
              "et": "On time",
              "length": "6",
              "locationName": "Hatfield",
              "st": "02:13",
            },
            {
              "crs": "WGC",
              "et": "On time",
              "length": "6",
              "locationName": "Welwyn Garden City",
              "st": "02:20",
            },
          ],
        },
      },
    ],
  },
}
`;
